### ISSUE TITLE
prefer-destructuring should support function parameter declaration

### ISSUE THREAD

**Issue by AndyRightNow** [eslint/eslint#12710]
**What rule do you want to change?**
prefer-destructuring
**Does this change cause the rule to produce more or fewer warnings?**
More warnings
**How will the change be implemented? (New option, new default behavior, etc.)?**
New option
**Please provide some example code that this change will affect:**

```js
function getFullName(user) {
  return `${user.firstName} ${user.lastName}`;
}
```

**What does the rule currently do for this code?**
It does not show warnings. See [this demo link](https://eslint.org/demo#eyJ0ZXh0IjoiZnVuY3Rpb24gZ2V0RnVsbE5hbWUodXNlcikge1xuICByZXR1cm4gYCR7dXNlci5maXJzdE5hbWV9ICR7dXNlci5sYXN0TmFtZX1gO1xufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OCwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXMiOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyLCJwcmVmZXItZGVzdHJ1Y3R1cmluZyI6Mn0sImVudiI6e319fQ==)
**What will the rule do after it's changed?**
Provide warnings at all usage of member expressions of `user` and it should be fixed to
```js
function getFullName({ firstName, lastName }) {
  return `${firstName} ${lastName}`;
}
```
**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by mdjermanovic** [comment]
Hi @AndyRightNow, thanks for the change request!

While this is also about destructuring, it's different.

`prefer-destructuring` just requires destructuring syntax in assignments and declarations, not creating a new variable for each property that is used in any way.

This looks to me more like something that could be a separate rule rather than an option in the `prefer-destructuring` rule.

> Provide warnings at all usage of member expressions of user and it should be fixed

What should the option (or a rule) check before warnings, e.g., should these be fixed:

```js
function getFullName(user) {
  user.firstName = "foo";
  return `${user.firstName} ${user.lastName}`;
}

function getFullName(user) {
  doSomething(user);
  return `${user.firstName} ${user.lastName}`;
}

function getFullName(user) {
  user.doSomething();
  return `${user.firstName} ${user.lastName}`;
}
```

**Comment by AndyRightNow** [comment]
Hi @mdjermanovic  Thank you for your reply.

To describe the rule in a more accurate way, it would probably be:

Prefer destructuring in function parameters if
1. There is no direct usage of or reference to the parameter identifiers
2. The parameter identifiers are not used as the context in member function calls
3. There is no AssignmentExpression with side effects


My rough idea is to add a "FunctionDeclaration" option. It would also be fine if a separate rule called something like "prefer-destructuring-in-unused-param-identifier".

In terms of the examples you provided, it should be

```js
function getFullName(user) {
  user.firstName = "foo";
  return `${user.firstName} ${user.lastName}`;
}
```
The AssignmentExpression has side effect (with MemberExpression in the left) so it does not apply.

```js
function getFullName(user) {
  doSomething(user);
  return `${user.firstName} ${user.lastName}`;
}
```
There is a direct usage of the parameter identifier in `doSomething(user);` so the rule does not apply.

```js
function getFullName(user) {
  user.doSomething();
  return `${user.firstName} ${user.lastName}`;
}
```
The parameter identifier here is used as the context in member function call.

A longer example that applies is like this:
```js
function getFullAddress(user) {
  user.utils.doSomething();

  return [
    `${user.firstName} ${user.lastName}`,
    user.addressLine1,
    user.addressLine2,
    user.addressLine3,
    `${user.city} ${getAbbr(user.state)} ${getAbbr(user.zipCode)}`,
    user.country
  ].filter(text => text)
    .join('\n');
}
```
and it will be fixed to

```js
function getFullAddress({ 
  firstName,
  lastName,
  addressLine1,
  addressLine2,
  addressLine3,
  city,
  state,
  zipCode,
  country,
  utils,
}) {
  utils.doSomething();

  return [
    `${firstName} ${lastName}`,
    addressLine1,
    addressLine2,
    addressLine3,
    `${city} ${getAbbr(state)} ${zipCode}`,
    country
  ].filter(text => text)
    .join('\n');
}
```

**Comment by mdjermanovic** [comment]
`user.utils.doSomething()`/`utils.doSomething()` could possibly modify `user`'s properties, so the fix might be wrong?

Actually, even this might change `user.firstName`/`user.lastName` in some cases:

```js
function getFullName(user) {
  doSomething();
  return `${user.firstName} ${user.lastName}`;
}
```

This is an interesting idea, but it doesn't seem it can be reliably enforced.

**Comment by AndyRightNow** [comment]
@mdjermanovic 

Got it. I was originally assuming this is done in a mostly functional programming context but operations with side effects could render the fixed code completely incorrect. 

I would probably implement it in a custom plugin to fit a more specific use case. I will close the issue then. Thank you for your time!
